src/main/scala/freeeventsourcing/syntax/ProcessSyntax.scala

```diff
@@ -0,0 +1,218 @@
+package freeeventsourcing.syntax
+
+import java.time.Instant
+import scala.annotation.implicitNotFound
+import cats.Monad
+import cats.free.Free
+import freeeventsourcing._
+import freeeventsourcing.EventSelector._
+import freeeventsourcing.ProcessAction.FirstOf.{ Alternative, Alternatives }
+import freeeventsourcing.ProcessAction._
+import freeeventsourcing.support.{ AggregateFromId, ValidAggregate }
+import freeeventsourcing.utils.StringSerializable
+import shapeless.ops.coproduct.{ Remove, Selector, Unifier }
+import shapeless.{ :+:, ::, CNil, Coproduct, HList, HNil, Inl, Inr }
+
+/** Nice monadic syntax to write processes. */
+class ProcessSyntax[BC <: BoundedContext](boundedContext: BC) {
+  type Action[+A] = ProcessAction[BC, A]
+  type ProcessMonad[A] = Free[Action, A]
+
+  type ValidSelector[S] = freeeventsourcing.support.ValidSelector[BC, S]
+
+  import Builders._
+
+  /** Await an event using a selector. The event must be from this bounded context. */
+  def await[S <: WithEventType: EventSelector: ValidSelector](selector: S) =
+    lift(AwaitEvent[BC, S](selector))
+
+  /** Send commands to a specific aggregate. */
+  def on[Id, A](aggregate: Id)(implicit afi: AggregateFromId[Id, BC#Aggregates]) = {
+    val aggregateType = afi.aggregate(boundedContext.aggregates)
+    new OnAggregateBuilder[afi.Out](aggregateType, aggregate)
+  }
+
+  /** Await events from a specific aggregate. */
+  def from[Id, A](aggregate: Id)(implicit afi: AggregateFromId[Id, BC#Aggregates]) = {
+    val aggregateType = afi.aggregate(boundedContext.aggregates)
+    new FromAggregateBuilder[afi.Out](aggregateType, aggregate)
+  }
+
+  /** Wait until the specified date/time. If the time has already passed it will still be called. */
+  def waitUntil(when: Instant) =
+    lift[Unit](WaitUntil[BC](when))
+
+  /** Wait for multiple events and run the path of the first event. */
+  def firstOf[Paths <: HList](b: FirstOfBuilder[HNil] ⇒ FirstOfBuilder[Paths])(implicit switch: Switch[Paths]): ProcessMonad[switch.Result] = {
+    val paths = b(new FirstOfBuilder(HNil)).collect
+    val alternatives = switch.alternatives(paths)
+    lift[alternatives.Events](FirstOf[BC, alternatives.type](alternatives))
+      .flatMap(e ⇒ switch.effectFor(paths)(e))
+  }
+
+  /** Wait for multiple events and run the path of the first event. */
+  def firstOfUnified[Paths <: HList, R <: Coproduct](b: FirstOfBuilder[HNil] ⇒ FirstOfBuilder[Paths])(implicit s: Switch.Aux[Paths, R], u: Unifier[R]) =
+    firstOf(b).map(r ⇒ u(r))
+
+  /** Terminate this process instance. */
+  def terminate = lift[Unit](End[BC]())
+
+  /** Does nothing. */
+  def noop = Monad[ProcessMonad].pure(())
+
+  /////////////////////////////////////////////////////
+
+  private def lift[A](action: ProcessAction[BC, A]) = Free.liftF[ProcessAction[BC, ?], A](action)
+
+  /** Internas used by the above functions. Need to be public but are not meant to be used directly. */
+  object Builders {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Object name does not match the regular expression '"[A-Z][A-Za-z]*"'](https://www.codacy.com/app/msiegenthaler/free-event-sourcing/file/2571398663/issues/source?bid=3298199&fileBranchId=3298202#l68)